### ISSUE TITLE
Too Many Files Error with AWS ALB

### ISSUE THREAD

**Issue by nkitiyakara** [eclipse/jetty.project#4836]
I've opened a support case with AWS about this as well, but it might be worth someone looking at it to see about hardening against misbehaving load balancers or other components.

We have an application that would occasionally get this error:

java.io.IOException: Too many open files
at sun.nio.ch.ServerSocketChannelImpl.accept0(Native Method)
at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:419)
at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:247)
at org.eclipse.jetty.server.ServerConnector.accept(ServerConnector.java:385)
at org.eclipse.jetty.server.AbstractConnector$Acceptor.run(AbstractConnector.java:701)
at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
at java.lang.Thread.run(Thread.java:748)

After a fair amount of experimentation, I isolated the issue to connections made by the AWS Application Load Balancer apparently not being closed properly.  Creating an ec2 instance (Ubuntu 18.04) with a public IP address and connecting directly to that address did not cause the problem, but accessing that same machine via the ALB did cause the problem.

I can work around this issue by using an nginx instance to decrypt/re-encrypt the requests forwarded by the ALB, but this doesn't seem elegant to me.  

Please find attached the source code for a small sample application that can demonstrate the issue (although you will need to install in AWS and set up the ALB and so forth) and an Apache JMeter test for forcing it to happen (server names will not be valid, but it shows the idea). 

[leaky-jmeter-test.zip](https://github.com/eclipse/jetty.project/files/4565974/leaky-jmeter-test.zip)
[connection-leak-example_source.zip](https://github.com/eclipse/jetty.project/files/4565976/connection-leak-example_source.zip)

**Comment by joakime** [comment]
What version of Jetty?
What version of Java?

**Comment by nkitiyakara** [comment]
Sorry, end of the day thinking.

Jetty version 9.4.28.  (Full details can be determined from the
build.gradle file.)
I tested against both Java 8 and Java 11.  Same results in both cases.

*Narti Kitiyakara* | ConnectDER
703-887-7640 <(703)+887-7640> | nkitiyakara@ConnectDER.com

**Comment by joakime** [comment]
What is your open files configuration on that EC2 node?
What is your Jetty configuration on that EC2 node?

**Comment by joakime** [comment]
The example project shows a fairly simple HTTP/2 server faced by Javalin, serving 1 endpoint (it outputs some HTML with some context details; time, remote-addr, protocol, scheme).
The jmeter test is highly specific to your specific AWS nodes and configuration, so we cannot use it locally.
The jmeter behavior / test plan isn't that remarkable, quite ordinary, and basic.
Configuring the jmeter plan to use a local Jetty instance shows no issues with connections.

We test all of the active mainline Jetty branches for performance daily, with different configurations, and on different JVMs.
Each configuration has multiple physical clients managing the request load against a single server.
We have yet to find a way for a single client machine to generate enough load to tax a single Jetty server on a default configuration.
We also haven't encountered a connection leak yet.

**Comment by nkitiyakara** [comment]
As I said, the connection leak only happened when I accessed the site through the AWS ALB, and I have a support case open with them, too.  My suspicion is that the ALB is doing something incorrectly that causes the connection leak, so I don't expect that you'd be able to reproduce it locally, anyway.  I just wanted you be aware of it in care there was some hardening that could be done on the Jetty side of things.

If it will help, I can gather some packet sniffing data to show what's happening from that perspective.  (Sorry, meant to have it ready initially, but my day ended up being more rushed than I expected.)

**Comment by nkitiyakara** [comment]
I reran my experiment for each access method (direct and through an ALB) with tcpdump running for part of each.  I left tcpdump running for around 20 minutes each time, but at that point the file was about 2.5 GB, so I didn't want to let tcpdump run longer and didn't want to attach the full files here.  (I'm uploading to Google drive and can send a shared link if you want the full capture.)  Instead, I've attached a capture filtered down to one TCP stream by Wireshark.  

It may be worth noticing that the connection termination from the ALB looks incorrect (based on my limited understanding from https://www.geeksforgeeks.org/tcp-connection-termination/).  When talking to the ALB, Jetty sends  the FIN flag, the ALB sends an ACK and then the ALB sends the RST,ACK flags.  I don't see the ALB sending its own FIN flag, which I thought was necessary.

When JMeter (running on Java 8, incidentally, Jetty was on Java 11) is talking directly to Jetty, Jetty sends the FIN and then JMeter responds with FIN,PSH,ACK all in one frame.  Jetty sends a final ACK and then JMeter sends several RST frames.  Wireshark calls the FIN,PSH,ACK frame suspicious, but Jetty does seem to let go of the connection in this case.

The attached txt files show what files Jetty had open at randomly selected times.  Each starts with the process limits and then I periodically ran "date && lsof -p $PID" for the appropriate process.  In the files-through-alb.txt file, you'll see that there were over 4000 files open by the Jetty process after about 1.25 hours.  I don't know if it shows in this file or not, but it wasn't a steady increase.  There were times that it dropped back, but the overall trend was upward.  (The actual application has a limit of 4096 files.  I wasn't thinking about it and ran this example application as root, which may account for it allowing so many more open files.)

I only ran the test connecting directly to Jetty for about 50 minutes, but you'll see in the files.direct.txt file that it stayed steady at around 250 open files after I started JMeter.  (In both files, the first output of lsof is before making any connections; after that it's whenever I thought to capture the data.  The number after the listing is the approximate number of files in the detailed listing.  It's just the result of running "lsof -p $PID | wc-l" immediately after the lsof that produced the listing.)

I'm still reasonably sure this is primarily an issue with the ALB behaving badly, but it does seem like a route for a denial of service attack, so I thought someone here might want to look at it.  This is really lower level than I normally deal with, though, so for all I know it may be an issue with the Java implementation itself.  Just in case, here's what Jetty was running on:

ubuntu@ip-10-1-0-112:~$ java -version
openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment (build 11.0.7+10-post-Ubuntu-2ubuntu218.04)
OpenJDK 64-Bit Server VM (build 11.0.7+10-post-Ubuntu-2ubuntu218.04, mixed mode, sharing)


[one-stream-each.zip](https://github.com/eclipse/jetty.project/files/4567996/one-stream-each.zip)
[files-through-alb.txt](https://github.com/eclipse/jetty.project/files/4567998/files-through-alb.txt)
[files-direct.txt](https://github.com/eclipse/jetty.project/files/4568000/files-direct.txt)

**Comment by nkitiyakara** [comment]
Just as an additional note, the AWS people say they're doing a half-duplex termination of the connection, which they say is legitimate.  (This is really lower-level than I normally deal with.)

I also converted the test application I was using to a WAR (source attached) and deployed it to Tomcat instead of using an embedded Jetty.  When running the same tests against the application running in Tomcat, I still see the half-duplex terminations, but there isn't an increasing number of files being left open.  (I do see a few hundred connections in TIME_WAIT state, but the number of them stabilizes after a few minutes.)  In particular, Jetty is opening a lot of pipes that don't seem to get closed, while Tomcat is not opening any pipes.

[tomcat-connection-leak-example.zip](https://github.com/eclipse/jetty.project/files/4617188/tomcat-connection-leak-example.zip)

**Comment by sbordet** [comment]
@nkitiyakara I looked at the pcap streams and they are different.

In the direct case, there is encrypted data immediately followed by a FIN, which is replied correctly.

In the ALB case, there is data flowing normally, then a 30 s pause, then Jetty emits a TLS close_notify + FIN which is replied with a RST.

This means the application is doing things differently; in the direct case it closes the connection explicitly, in the ALB case it is not closed by the application, and it's closed by Jetty's idle timeout, to which the ALB replies with a RST.

You want to be sure that the ALB times out before Jetty since it's always good to have clients idle timing out connections before the server.

I believe we do have a test case where the client sends a RST, I'll dig it up or write one. 
I will let know if we handle this case well (or not).

**Comment by sbordet** [comment]
@nkitiyakara I have been able to replicate exactly the traffic you see from the ALB (idle timeout + FIN + RST), and confirmed that Jetty does close the server EndPoint as it should.

If you can replicate easily, next step is to enable DEBUG logging in Jetty and attach them here.
So far, seems all good to us.

**Comment by nkitiyakara** [comment]
About 6 minutes worth of log file was 56 MB zipped, so I've shared it from Google Drive again (https://drive.google.com/file/d/1r3H19f2FGBWEDbmrM1xAbui4U5DG0vhS/view?usp=sharing).  Also included in the ZIP file are txt files showing the output of "sudo lsof -c java" and "sudo netstat -anp" at about 30 second intervals.  The problem I'm experiencing is exposed better in the lsof output (files.txt).  If you count the number of pipes in each run, you'll see it growing.  It would eventually exceed the number of open files allowed for the process.  (This run was with 50 JMeter users with a 100 second ramp up period, incidentally.)

**Comment by sbordet** [comment]
@nkitiyakara sorry but again all seems good to me.

Your `ports.txt` file shows that at the end, all TCP ports have been closed.
It shows, just before the end, 135 connections opened to port `8443`.
It's a small number, so if your server chokes on that, you have bigger configuration problems.

I picked a couple of random ports from those 135 (`37264` and `37204`) and I can see that they are correctly closed by Jetty. I have not run through them all, but I think they are all closed properly.

Did you have an answer from the ALB support?

**Comment by nkitiyakara** [comment]
It's actually file handles that have been running out rather than TCP ports.  (I don't know why Jetty is opening pipes, but it's those that don't seem to get closed.)  You can see the effect in the files-through-alb.txt attachment to my message of 2 May.

The AWS people said they were doing a half-duplex close, which they say is valid, although Jetty does seem to close those pipe when there is a full-duplex close (as when I have JMeter hit a public IP address of the test system without going through an ALB).

**Comment by sbordet** [comment]
Jetty does not open pipes, so I'm not sure what you are seeing.

IIRC everything is fine without ALB, right?

**Comment by nkitiyakara** [comment]
Yep, everything is fine without the ALB in the mix.

That is the PID for the process running Jetty, but it could be Consypt or Javalin opening the pipes, too.  Those were the only other components I used for the test application.  (The code I wrote directly for the test deliberately didn't use anything beyond information from the HttpServletRequest and a LocalDateTime, so it's hard to see that causing any problems.)

Thanks for looking into this.
